### PR TITLE
Ensure `Snackbar` + `Toast` work in Modal Page

### LINKED ISSUE #334: [Bug] Snackbar isn't shown in a modal page
### Description

The snackbar does not show when invoked from within a modal page.

### Steps to Reproduce

1. Create a new Maui project
2. Add a new page `Page2`
3. From within the main page add a button and in the click handler: `Navigation.PushModalAsync(new Page2());`
4. In `Page2` add a button and in the click handler: `await Snackbar.Make("test").Show();`

### Expected Behavior

The snackbar should show.

### Actual Behavior

The snackbar does not show.

### Basic Information

- Version with issue: 1.0.0-pre8
  - Android: 12.1

### LINKED ISSUE #344: [Bug] iOS SnackBar Message Doesn't Have Action Button
### Description
I've attached a sample project that demonstrates the issue.
The snackbar message doesn't show the action button in iOS. The action button for the snackbar does
show in Android.
As an additional note changing font size doesn't seem to affect this.

### Steps to Reproduce
In iOS
1. Run sample project.
2. Click "Open Snack Bar Message"
3. The snackbar message appears without an action OK button.

### Expected Behavior
The snackbar message should have an action button in iOS.

### Actual Behavior
The snackbar message does not have an action button in iOS.

### Basic Information
- Version with issue: CommunityToolkit.Maui 1.0.0-pre8, Maui 14
[MauiFlyout.zip](https://github.com/CommunityToolkit/Maui/files/8383946/MauiFlyout.zip)

### PR DESCRIPTION
**Pull request by VladislavAntonyuk**
### Description of Change ###

 

 ### Linked Issues ###
 

 - Fixes #334 
 - Fixes #344 

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls/XYZ 

 ### Additional information ###

### CODE REVIEW

File: src/CommunityToolkit.Maui/Alerts/Toast/Toast.android.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/Alerts/Toast/Toast.android.cs:28]: Why can't we use the `AppContext`?

Can we avoid the `Native` word? Maybe we can use `PlatformToast`

**Review by brminnick — APPROVED** [review]
Thanks Vlad!!

✅ Confirmed Snackbar appears on modal pages on iOS, Android + macOS
✅ Confirmed Toast appears on modal pages on iOS, Android + macOS

**Comment by brminnick** [comment]
Dang - looks like the Unit Tests are complaining about `MainThread` not being implemented on `net6.0`.

You'll probably just have to create a mock for `IMainThread`. I had to do something similar for my app, GitTrends: https://github.com/brminnick/GitTrends/blob/main/GitTrends.UnitTests/MockServices/MockMainThread.cs

File: src/CommunityToolkit.Maui/Alerts/Toast/Toast.android.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Alerts/Toast/Toast.android.cs:28]: Because we need a context of the RootView, while the AppContext is Application context

**Comment by VladislavAntonyuk** [comment]
> 

IMainThread doesn't exist. MainThread is a static class

**Review by pictos — COMMENTED** [review]
@VladislavAntonyuk thanks for renaming everything <3.
For me, this is ready to merge, but I'll wait for @brminnick approval since there are issues with the unit test.

I know that the MAUI team is refactoring the essentials to use interfaces, we can wait on that... So, should we merge this as it's and create a ticket to improve the unittest?

**Comment by VladislavAntonyuk** [comment]
@pictos I try to replace MainThread with Dispatcher.
Do we need MainThread.IsMainThread if we have Dispatcher.GetForCurrentThread()? If I am not mistaken Dispatcher ensures that we run on UI thread

**Comment by VladislavAntonyuk** [comment]
We have some natives in popup, but it is for a separate PR

**Comment by pictos** [comment]
@VladislavAntonyuk 

> @pictos I try to replace MainThread with Dispatcher. Do we need MainThread.IsMainThread if we have Dispatcher.GetForCurrentThread()? If I am not mistaken Dispatcher ensures that we run on UI thread

So, the `Dispacher` should work, since it's in a `shared` level and will take care of platform specifics. But I would suggest to keep the Essentials APIs, just because MAUI will rely on this API.



> We have some natives in popup, but it is for a separate PR

Yeap, I'll take a look at those when I've time. Rush week for me

**Comment by VladislavAntonyuk** [comment]
MainThread will be moved to Compatibility with the next release, so we can replace it with Dispatcher: https://github.com/dotnet/maui/pull/5407#issuecomment-1076965497

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.shared.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.shared.cs:173]: This code doesn't run on `net6.0`, only on `net6.0-ios`, `net6.0-android`, `net6.0-maccatalyst`, and `net6.0-windows`, so it shouldn't impact our Unit Tests.

File: src/CommunityToolkit.Maui/Alerts/Toast/Toast.shared.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui/Alerts/Toast/Toast.shared.cs:124]: Can we keep this, replacing `MainThread` with `Dispatcher`? 

We added this after getting random crashes in the sample app that contained cryptic error messages. 

The crashes occurred because `Snackbar` was being modified from a non-UI thread, which makes sense. However, the StackTrace / Error Message provided from OS was very vague and unclear about _why_ the crash happened. 

Long story short, we added this `MainThread.IsMainThread` logic so that when future devs try to consume the `Snackbar` from a background thread will get a clear + concise error message that tells them exactly why it crashed, rather than some confusing OS exception like `SIGSEV blahblahblah`.

Comment (brminnick) [reply src/CommunityToolkit.Maui/Alerts/Toast/Toast.shared.cs:124]: This code doesn't run on `net6.0`, only on `net6.0-ios`, `net6.0-android`, `net6.0-maccatalyst`, and `net6.0-windows`, so it shouldn't impact our Unit Tests.

**Review by brminnick — CHANGES_REQUESTED** [review]
Thanks Vlad!!

I'd like to keep the `IsMainThread` check (moving it to use `Dispatcher`) if we can! I added more details on my comment 👍

File: samples/CommunityToolkit.Maui.Sample/Platforms/Windows/App.xaml.cs
Comment (VladislavAntonyuk) [reply samples/CommunityToolkit.Maui.Sample/Platforms/Windows/App.xaml.cs:38]: It's for the Snackbar action to make sure we have only 1 application instance. (Before it opened a new window).

Comment (VladislavAntonyuk) [reply samples/CommunityToolkit.Maui.Sample/Platforms/Windows/App.xaml.cs:38]: We need static field mutex, otherwise it's disposed and cleaned by GC

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.shared.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.shared.cs:173]: MainThread as I left a comment is going to be removed. Also according to implementation Dispatcher checks if it needs to be dispatched and run the code on platform MainThread.
Because our methods Show, Dismiss, Dispose use Dispatcher (and it's the only one way to show Snackbar/Toast) from my point of view this check is redundant.

File: samples/CommunityToolkit.Maui.Sample/Platforms/Windows/App.xaml.cs
Comment (pictos) [reply samples/CommunityToolkit.Maui.Sample/Platforms/Windows/App.xaml.cs:38]: But we create a new mutex when this method is called and after that we just use the out parameter, so do you know why gc collecting the mutex object is bad?

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.windows.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.windows.cs:69]: While ago we discussed forcing the code to execute on `MainThread` inside the lib, and how that isn't a good practice. One reason is that this can cause some bugs and then could be hard to identify. For example, here, the `Action` will not be executed right after you step from this method, it will be queued to be executed when the `MainThread` is free, so at the time that this `Action` execute the state of the application can be invalid and this can lead to some crash or odd behavior. 

From the usage of this, this `delegate handler` will be subscribed in the thread that calls `ShowPlatform`, and the **user** must call it from the `MainThread`. What we can do here to protect the user is adding a check to make sure that we're on the `MainThread` and throw a good exception message. More than that I would say that we're shooting ourselves in the foot.

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.windows.cs:69]: It doesn't work when you use MainThread class: image

File: samples/CommunityToolkit.Maui.Sample/Platforms/Windows/App.xaml.cs
Comment (VladislavAntonyuk) [reply samples/CommunityToolkit.Maui.Sample/Platforms/Windows/App.xaml.cs]: updated

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.windows.cs
Comment (pictos) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.windows.cs:69]: What I mentioned is that we shouldn't use either `Dispatcher` or `MainThread`.
we should just do `Action();` or `Action.Invoke()`

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.shared.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.shared.cs:173]: Yup, you're right that it's totally redundant because `Show()` and `Dismiss()` always push to the UI thread. 

I added the redundancy on purpose for two reasons:
- If `Show()` and `Dismiss()` are ever modified to remove the Dispatcher, we will still validate that we're using the correct thread
- Since `Snackbar` is a partial class, if anything in `Snackbar.[platorm].cs` references `Snackbar`, this ensures `NativeSnackbar` is always referenced from the UI thread

All that said, this isn't really something I care enough about to hold up this PR! If you'd like to keep it out, that's cool. Just let me know and I'll approve this PR 👍

Comment (brminnick) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.shared.cs:173]: Fair enough. We can refactor `Show()` and `Dismiss()` to remove the marshaling to the UI thread.

I strongly disagree that a library should provide code that throws a preventable exception, but I'm happy to be outvoted by Vlad, Pedro and Shaun 🙌

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.windows.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.windows.cs:69]: it is important to keep Dispatcher here:
image

**Review by brminnick — APPROVED** [review]
Thanks Vlad! Just FYI - now that we don't need to `await` anything for the `Dispose` workflow, I changed `IAlert : IAsyncDisposable` to `IAlert : IDisposable`

File: src/CommunityToolkit.Maui/Alerts/Toast/Toast.shared.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Alerts/Toast/Toast.shared.cs:121]: isDisposed=true

**Review by brminnick — APPROVED** [review]

File: src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs:11]: In `Maui/Forms` isn't common to have don't allow inheritance on controls, do you have a good motivation for that?

File: src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs:13]:
```suggestion
	readonly List<UIView> children = Enumerable.Empty<UIView>().ToList();
```

This will be faster

```
|               Method |     Mean |    Error |   StdDev | Ratio |  Gen 0 | Allocated |
|--------------------- |---------:|---------:|---------:|------:|-------:|----------:|
|       UsingArrayEmpty | 17.36 ns | 0.164 ns | 0.145 ns |  1.00 | 0.0051 |      32 B |
| UsingEnumerableEmpty | 10.22 ns | 0.060 ns | 0.047 ns |  0.59 | 0.0051 |      32 B |
```

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.macios.cs
Comment (pictos) [inline src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.macios.cs:36]: I see here that we use the `DefaultPadding`, but what happens if the user changes it?

**Review by pictos — COMMENTED** [review]
@VladislavAntonyuk I have some questions and suggestions before merging this one.

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.macios.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.macios.cs:36]: The user will get different padding. 
It is a public setting.

Comment (pictos) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.macios.cs:36]: Default padding is a read-only and we don't set it inside the `PlatformToast`[ ctor](https://github.com/CommunityToolkit/Maui/blob/eb2606b34d56b96e8cb71eb0dd9f923ffa2d4e3a/src/CommunityToolkit.Maui.Core/Views/Toast/PlatformToast.macios.cs#L35-L50). So even if the user sets a padding on ctor, the DefaultPadding will be the same

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.macios.cs:36]: Ah, you are right. I need a cup of coffee. I thought it is in Defaults.shared.cs.
I don't remember why it is a public read only. It should be private constant. We need this property to look similar to Android Snackbar.

File: src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs:11]: Inspired by https://www.meziantou.net/performance-benefits-of-sealed-class.htm

Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs:11]: Yeah, I'm very cautious with inheritance as well. But in this case, we should allow it and we're prepared for it.

File: src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.macios.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Alerts/Snackbar/Snackbar.macios.cs]: so to summarize, it is used as a default padding. so if the user sets the corner radius equal to 0, we still have some padding equal to 10. but it was not correct to set it here. Now set default padding internally

**Comment by VladislavAntonyuk** [comment]
Added additional documentation PR: https://github.com/MicrosoftDocs/CommunityToolkit/pull/55

**Review by brminnick — APPROVED** [review]

**Review by pictos — APPROVED** [review]

**Pull request merged**
